### ISSUE TITLE
Dashboard generator misbehaves with translated content

### ISSUE THREAD

**Issue by sedubois** [thoughtbot/administrate#1668]
### What would you like to be able to do? Can you provide some examples?

I started adding Administrate to my bilingual app when I was also using Globalize. In terms of modelling, Globalize stores the translations on a separate table, but conceptually the attributes are those of the parent model. However Administrate is not aware of this so I ended up in a situation when I had two dashboards and two controllers (with [some workarounds](https://github.com/thoughtbot/administrate/issues/1291#issuecomment-545054789) because Administrate has some bugs with namespaced models).

So this whole time I was using this working, albeit complicated system where every model's data was split between two dashboards: one for untranslated attributes and one for translated attributes. The translated attributes were linked from the base dashboard using a NestedHasMany field, which got the job done (`translations: Field::NestedHasMany.with_options(class_name: "Post::Translation")`).

However now that I want to migrate from Globalize to Mobility with a key-value i18n backend, it made me realize that Administrate should in fact not generate dashboards this way, but keep the translated attributes in the main dashboard. The fact that those attributes are stored in a separate table, in a polymorphic key-value table, in a JSONB column, or somewhere else, is an "implementation detail" which shouldn't impact the way the dashboards are structured and the API gets queried.

Now I've re-combined all the dashboards together and the dashboard definitions became much cleaner. 

As now the different translations don't show up at the same time any more, I've added a language switcher to the navigation. I use `route_translator` and added this to the navigation:

<details>

```erb
<% (I18n.available_locales - [I18n.locale]).each do |locale| %>
  <% I18n.with_locale(locale) do %>
    <% if @requested_resource.try(:to_param) %>
      <% url_params = [:admin, @requested_resource, **request.query_parameters] %>
    <% else %>
      <% url_params = { path_locale: locale.to_s, **request.query_parameters } %>
    <% end %>
    <%= link_to t("common.language"), url_for(url_params), class: "button button--alt" %>
  <% end %>
<% end %>
```

</details>

(One thing I have not yet figured out, is how to ensure that the search bar works again. Now for example if I have a translated `title` attribute, it throws:)
```PG::UndefinedColumn: ERROR:  column posts.title does not exist
LINE 1: ... AS CHAR(256))) LIKE '%hello%' OR LOWER(CAST("posts"."t...
```

Anyway, so it would be great if Administrate somehow recognized that the concerned attributes are translation attributes and kept them together with the untranslated ones when generating the dashboards, instead of ending up with two dashboards and two controllers.

I am opening this ticket both to document these "realizations" in the hope it helps someone else, and also as a feature request in case it would be possible for Administrate to generate the dashboards correctly. I don't really need the feature myself any more but others might be thankful not to encounter the above described behaviour.

### How could we go about implementing that?

Very good question. How can this be done in a way that is independent enough from the underlying translation implementation?

As with ActionText `has_rich_text`, translations can be stored in a key-value table linked by a polymorphic relation. (This is what I actually aim to end up with my migrating to Mobility.) So whatever solution works for ActionText might work for this as well.

### Can you think of other approaches to the problem?

Maybe just add a warning somewhere like "if you use a translation backend, make sure to fix your dashboards after they have been generated by doing this and that".

**Comment by pablobm** [comment]
@sedubois - Would you be able to create a simple example app, with the issue you describe, that we can play with? I don't quite understand this 100%, and I don't know even where to start from.

**Comment by sedubois** [comment]
@pablobm 

```
rails new administrate_mobility_demo && cd $!
rails g scaffold Post title:string content:text status:integer
bundle add administrate mobility
rails generate mobility:install
ruby -i -pe '$_.gsub!(/end/, "  extend Mobility\n  translates :title, type: :string\n  translates :content, type: :text\nend")' app/models/post.rb
rails db:migrate
rails g administrate:install
rails server
```

At this point I hit an error `Asset `administrate/application.js` was not declared to be precompiled in production.` which I don't know how to fix (adding `//= link administrate/application.css` to `app/assets/config/manifest.js` and restarting server doesn't help).

Current status is here:
https://github.com/sedubois/administrate_mobility_demo

**Comment by sedubois** [comment]
@pablobm I've updated previous comment with more explanations.

Also I made another demo with Globalize that shows the original issue I had:

```
rails new administrate_globalize_demo && cd $!
rails g scaffold Post title:string content:text status:integer
bundle add administrate globalize
ruby -i -pe '$_.gsub!(/end/, "  translates :title, :content\nend")' app/models/post.rb
cat > db/migrate/20200707092509_create_post_translations.rb << EOF
class CreatePostTranslations < ActiveRecord::Migration[6.0]
  def up
    Post.create_translation_table!(
      title: :string,
      content: :text
    )
  end

  def down
    Post.drop_translation_table!
  end
end
EOF
rails db:migrate
rails generate administrate:install
rails server
```

Demo: https://github.com/sedubois/administrate_globalize_demo

Even though it has the same (unrelated) Sprockets issue as in previous comment, you can see that the PostDasboard in this case was not generated with the `title` and `content`, they are missing:
```rb
class PostDashboard < Administrate::BaseDashboard
  # ...
  ATTRIBUTE_TYPES = {
    translations: Field::HasMany.with_options(class_name: "Post::Translation"),
    id: Field::Number,
    status: Field::Number,
    created_at: Field::DateTime,
    updated_at: Field::DateTime,
  }.freeze
  #....
```

Furthermore, there is the same issue as with Mobility that it also generates a `translations` which is included in the `COLLECTION_ATTRIBUTES`, `SHOW_PAGE_ATTRIBUTES` and `FORM_ATTRIBUTES`.

So to summarize:
- With Globalize the dashboard generation is broken as the translated attributes are missing. However Globalize is no [longer maintained](https://github.com/globalize/globalize/issues/753) and users are encouraged to move to Mobility, so we could ignore this issue.
- With both Mobility and Globalize, the generated dashoard gets `*translations` attributes but these should be avoided if possible (no idea how Administrate would figure that out, though).

**Comment by pablobm** [comment]
Thank you so much for the comprehensive info and step-by-step guides @sedubois. It's been really helpful.

I have created `administrate_mobility_demo` as per your instructions. The first thing I noticed is that I don't have that problem with the assets manifest, after I add the entries for the Administrate assets. In your comment you mention the line for CSS. Did you add the line for JS too?

Going into the actual issue that you describe, it appears that Mobility's `translate` macro generates a `has_many` relationship for each attribute. This leaves things in a state where Administrate can't tell if the association is "legit" or not. In other words: it's unlikely that there's a way for Administrate to tell those `has_many` relationships from other, "normal" `has_many` relationships. Therefore, there's nothing that the generator can do. It's up to the user to correct those dashboards manually.

I haven't looked into Globalize, but I assume it's the same thing: the tool provides a translation interface on top of Rails's own lower-level constructs. When Administrate's generator inspects the models, there's nothing in there that indicates that anything is out of the ordinary.

This is what the dashboard generator currently does to figure out the attributes in a model:

https://github.com/thoughtbot/administrate/blob/master/lib/generators/administrate/dashboard/dashboard_generator.rb#L54-L58

All that the generator can do is look at the model and figure out from there.

**Comment by pablobm** [comment]
I think this is out of scope, but perhaps it could be added to the documentation. I'm closing the issue, but adding it to the list of things that could go in the guides.

**Comment by dreyks** [comment]
FWIW, this is my administrate+globalize setup

- let's say i have a model with a translated "name" column
- i'm using globalize-accessors gem that gives me `name_en` and `name_en=` methods
- in my dashboard's ATTRIBUTES i've added `name_en: Field::String.with_options(searchable: false)`
- i also added `name_en` to FORM_ATTRIBUTES so that i can update the value (administrate doesn't need to know that it goes to a different table)
- to make the name column actually searchable i've added another attribute `translations: Field::HasMany.with_options(searchable: true, searchable_fields: [:name])` that is not present in any other _ATTRIBUTES array
